src/VisualStudio/Core/Def/Implementation/FindReferences/AsyncFindReferencesPresenter.DataSource.cs

```diff
@@ -0,0 +1,180 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editor;
+using Microsoft.CodeAnalysis.Editor.Navigation;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem;
+using Microsoft.VisualStudio.Shell.TableManager;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.FindReferences
+{
+    internal partial class AsyncFindReferencesPresenter
+    {
+        private class DataSource : FindReferencesContext, ITableDataSource, ITableEntriesSnapshotFactory
+        {
+            private readonly CancellationTokenSource _cancellationTokenSource = new CancellationTokenSource();
+            private readonly ConcurrentBag<Subscription> _subscriptions = new ConcurrentBag<Subscription>();
+            private readonly IAsynchronousOperationListener _asyncListener;
+
+            private readonly object _gate = new object();
+            private ImmutableArray<TableEntry> _entries = ImmutableArray<TableEntry>.Empty;
+
+            private TableEntriesSnapshot _lastSnapshot;
+
+            public DataSource(IAsynchronousOperationListener asyncListener)
+            {
+                _asyncListener = asyncListener;
+            }
+
+            public override CancellationToken CancellationToken => _cancellationTokenSource.Token;
+
+            #region ITableDataSource
+
+            public string DisplayName => "Roslyn Data Source";
+
+            public string Identifier => "Roslyn Identifier";
+
+            public string SourceTypeIdentifier => "FindAllReferencesProvider";
+
+            public int CurrentVersionNumber { get; private set; }
+
+            public IDisposable Subscribe(ITableDataSink sink)
+            {
+                var subscription = new Subscription(_cancellationTokenSource, sink);
+                _subscriptions.Add(subscription);
+
+                sink.AddFactory(this, removeAllFactories: true);
+
+                return subscription;
+            }
+
+            #endregion
+
+            #region EditorFindReferencesContext overrides.
+
+            public override void OnStarted()
+            {
+                foreach (var subscription in _subscriptions)
+                {
+                    subscription.TableDataSink.IsStable = false;
+                }
+            }
+
+            public override void OnCompleted()
+            {
+                foreach (var subscription in _subscriptions)
+                {
+                    subscription.TableDataSink.IsStable = true;
+                }
+            }
+
+            public async override void OnReferenceFound(INavigableItem definition, INavigableItem reference)
+            {
+                var workspace = reference.Document.Project.Solution.Workspace as VisualStudioWorkspaceImpl;
```

Comment:
Probably yes to both.  Do those workspaces have ways to get the project guid for a Workspace Project?  That's basically why i need this code.  the new FAR window needs that project guid.